### PR TITLE
Replace accessControlHandler with authorization interceptor

### PR DESCRIPTION
**Pull request by sergeybykov**
**What changed?**
Removed `accessControlHandler` and replaced it with a `authorization.Interceptor`.

**Why?**
The interceptor is generic, doesn't need to the implement workflow service interface, and is invoked before the workflow handler. This allows authorization logic to execute first, to make sure only authorized calls are allowed to get through.

**How did you test it?**
Unit tests and integration tests.

**Potential risks**
I don't see any real risk as by default no authorization is done yet.

### CODE REVIEW

File: common/authorization/authorizationInterceptor.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	"go.temporal.io/server/common/metrics"
+)
+
+var (
+	errUnauthorized = serviceerror.NewPermissionDenied("Request unauthorized.")
+)
+
+func (a *interceptor) Interceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler,
+) (interface{}, error) {
+
```
Comment (wxing1292) [inline common/authorization/authorizationInterceptor.go:43]:
```go
func (a *interceptor) Interceptor(
    ctx context.Context, 
    req interface{}, info *grpc.UnaryServerInfo, 
    handler grpc.UnaryHandler,
) (interface{}, error) {
```

File: common/authorization/authorizationInterceptor.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	var ns string
```
Comment (wxing1292) [inline common/authorization/authorizationInterceptor.go:44]: ns -> namespace

File: common/authorization/authorizationInterceptor.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	rwns, ok := req.(RequestWithNamespace)
```
Comment (wxing1292) [inline common/authorization/authorizationInterceptor.go:45]: rwns -> requestWithNamespace

File: common/authorization/authorizationInterceptor.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	if ok {
+		ns = rwns.GetNamespace()
+	}
+
+	apiName := info.FullMethod
+	index := strings.LastIndex(apiName, "/")
+	if index > -1 {
```
Comment (wxing1292) [inline common/authorization/authorizationInterceptor.go:52]: do we expect if info.FullMethod does not contain a `/` ?

File: common/authorization/authorizationInterceptor.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+	}
+	return handler(ctx, req)
+}
+
+// checks if this is one of the four namespace operations
+func isNamespaceOp(api string) bool {
+	return api == "RegisterNamespace" ||
+		api == "DescribeNamespace" ||
+		api == "UpdateNamespace" ||
+		api == "DeprecateNamespace"
```
Comment (wxing1292) [inline common/authorization/authorizationInterceptor.go:84]: could you define these APIs at the beginning of this file?
```go
var namespaceAPIs = map[string]struct{}{
    "RegisterNamespace": struct{},
    "DescribeNamespace": struct{},
    "UpdateNamespace": struct{},
    "DeprecateNamespace": struct{},
}
```

here do 
```go
_, ok := namespaceAPIs[api]
return ok
```

File: common/authorization/authorizationInterceptor.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+	"google.golang.org/grpc"
+
+	"go.temporal.io/server/common/metrics"
+)
+
+var (
+	errUnauthorized = serviceerror.NewPermissionDenied("Request unauthorized.")
+)
+
+func (a *interceptor) Interceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler,
```
Comment (wxing1292) [inline common/authorization/authorizationInterceptor.go:41]: unit test for this function?

File: common/authorization/authorizationInterceptor.go (hunk 4 above)
Comment (sergeybykov) [reply common/authorization/authorizationInterceptor.go:52]: They come in the of `/temporal.api.workflowservice.v1.WorkflowService/DescribeNamespace`. So a slash is 100% expected. Even if somehow it's not there, we'll use the full string as the API name.

**Review by alexshtin — APPROVED** [review]

File: common/authorization/authorizationInterceptor.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
+// The MIT License
```
[the remark this answers is not in the document]
Comment (sergeybykov) [reply common/authorization/authorizationInterceptor.go]: I already renamed it to interceptor.go. No spaces for snakes left. 🙂

**Review by wxing1292 — APPROVED** [review]

**Pull request merged**